src/librustc_typeck/outlives/implicit_infer.rs

```diff
@@ -0,0 +1,257 @@
+    while changed {
+        changed = false;
+        let mut visitor = InferVisitor {
+            tcx: tcx,
+            inferred_outlives_map: &mut inferred_outlives_map,
+            changed: changed,
+        };
+
+        //iterate over all the crates
+        tcx.hir.krate().visit_all_item_likes(&mut visitor);
+    }
+}
+
+pub struct InferVisitor<'cx, 'tcx: 'cx> {
+    tcx: TyCtxt<'cx, 'tcx, 'tcx>,
+    inferred_outlives_map: &'cx mut FxHashMap<DefId, Rc<Vec<ty::Predicate<'tcx>>>>,
+    changed: bool,
+}
+
+impl<'cx, 'tcx> ItemLikeVisitor<'tcx> for InferVisitor<'cx, 'tcx> {
+    fn visit_item(&mut self, item: &hir::Item) {
+        let def_id = self.tcx.hir.local_def_id(item.id);
+
+        let node_id = self.tcx
+            .hir
+            .as_local_node_id(def_id)
+            .expect("expected local def-id");
+        let item = match self.tcx.hir.get(node_id) {
+            hir::map::NodeItem(item) => item,
+            _ => bug!(),
+        };
+
+        let mut local_required_predicates = FxHashMap();
+        match item.node {
+            hir::ItemUnion(..) |
+            hir::ItemEnum(..) |
+            hir::ItemStruct(..) => {
+                let adt_def = self.tcx.adt_def(def_id);
+                for field_def in adt_def.all_fields() {
+                    let required_predicates: Vec<_> = required_predicates_to_be_wf(
+                        self.tcx,
+                        field_def,
+                        &local_required_predicates,
+                     ).into_iter().collect();
+                    local_required_predicates.insert(
+                        field_def.did,
+                        Rc::new(required_predicates),
+                    );
+                }
+            }
+            _ => {}
+        };
+
+        if local_required_predicates.len() > 0 {
+            self.changed = true;
+            self.inferred_outlives_map
+                .extend(local_required_predicates);
+        }
+    }
+
+    fn visit_trait_item(&mut self, trait_item: &'tcx hir::TraitItem) {}
+
+    fn visit_impl_item(&mut self, impl_item: &'tcx hir::ImplItem) {}
+}
+
+fn required_predicates_to_be_wf<'tcx>(
+    tcx: TyCtxt<'_, 'tcx, 'tcx>,
+    field_def: &ty::FieldDef,
+    infered_outlives_map: &FxHashMap<DefId, Rc<Vec<ty::Predicate<'tcx>>>>,
+) -> HashSet<ty::Predicate<'tcx>> {
+
+    let mut predicates = HashSet::new();
+    // Get the type of the field with identity substs applied.  For
+    // now, let's just see if that causes horrible cycles.
+    let field_ty = tcx.type_of(field_def.did);
+    match field_ty.sty {
+        // For each type `&'a T`, we require `T: 'a`
+        //
+        // If you have A = &'a T, then mt.ty here represents T and
+        // ty represents A (and r represents 'a). So we want T: 'a.
+        ty::TyRef(r, mt) => {
+            let outlives_pred = ty::Binder(ty::OutlivesPredicate(mt.ty, r))
+                .to_predicate();
+            predicates.insert(outlives_pred);
```

Comment:
We are going to want to recurse to check `mt.ty` and extract predicates from it as well.